crates/sui-move-build/src/linters/freeze_wrapped.rs

```diff
@@ -0,0 +1,259 @@
+                                );
+                                let mut d = diag!(
+                                    FREEZE_KEY_DIAG,
+                                    (fun.arguments.exp.loc, msg),
+                                    (sloc, uid_msg)
+                                );
+
+                                if nested {
+                                    d.add_secondary_label((
+                                        nested_sname.loc(),
+                                        "Indirectly wrapped object is of this type",
+                                    ));
+                                }
+                                env.add_diag(d);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        E::Builtin(_, e) => visit_exp(e, env, modules),
+        E::Freeze(e) => visit_exp(e, env, modules),
+        E::Vector(_, _, _, e) => visit_exp(e, env, modules),
+        E::Dereference(e) => visit_exp(e, env, modules),
+        E::UnaryExp(_, e) => visit_exp(e, env, modules),
+        E::BinopExp(e1, _, e2) => {
+            visit_exp(e1, env, modules);
+            visit_exp(e2, env, modules);
+        }
+        E::Pack(_, _, fields) => fields
+            .iter()
+            .for_each(|(_, _, e)| visit_exp(e, env, modules)),
+        E::ExpList(list) => {
+            for l in list {
+                match l {
+                    H::ExpListItem::Single(e, _) => visit_exp(e, env, modules),
+                    H::ExpListItem::Splat(_, e, _) => visit_exp(e, env, modules),
+                }
+            }
+        }
+        E::Borrow(_, e, _) => visit_exp(e, env, modules),
+        E::Cast(e, _) => visit_exp(e, env, modules),
+        E::Unit { .. }
+        | E::Value(_)
+        | E::Move { .. }
+        | E::Copy { .. }
+        | E::Constant(..)
+        | E::BorrowLocal(..)
+        | E::Unreachable
+        | E::Spec(..)
+        | E::UnresolvedError => (),
+    }
+}
+
+fn struct_fields<'a>(
+    sname: &P::StructName,
+    mident: &E::ModuleIdent,
+    modules: &'a UniqueMap<E::ModuleIdent, H::ModuleDefinition>,
+) -> Option<(&'a Vec<(P::Field, H::BaseType)>, Loc)> {
+    if let Some(mdef) = modules.get(mident) {
+        if let Some(sdef) = mdef.structs.get(sname) {
+            if let H::StructFields::Defined(sfields) = &sdef.fields {
+                // unwrap is safe since we know that mdef.structs.get succeeded
+                return Some((sfields, *mdef.structs.get_loc(sname).unwrap()));
+            }
+        }
+    }
+    None
+}
+
+fn contains_key(
+    sp!(_, bt): &H::BaseType,
+    modules: &UniqueMap<E::ModuleIdent, H::ModuleDefinition>,
+    field_depth: usize,
+) -> Option<(P::StructName, bool)> {
+    let H::BaseType_::Apply(ability_set, tname ,_ ) = bt else {
+        return None;
+    };
+
+    let sp!(_, tname) = tname;
+    if let H::TypeName_::ModuleType(mident, sname) = tname {
```

Comment:
Does this cover the case where you try to freeze an object like 

```rust
struct Wrapper<K: key> has key {
  id: UID,
  field: K,
}
```

Reply:
+1 for this, you should just be able to check that the field has key directly?